### ISSUE TITLE
OutOfMemoryError thrown from the UncaughtExceptionHandler at startup

### ISSUE THREAD

**Issue by loren** [elastic/elasticsearch#20302]
**Elasticsearch version**: 5.0.0.alpha5
**Plugins installed**: [none]
**JVM version**: 
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)

**OS version**:
Mac 10.11.6 & CoreOS

**Description of the problem including expected versus actual behavior**:
Elasticsearch seems to throw an OOM error at startup whenever unicast hosts cannot be resolved. I would expect it to not OOM, and ideally behave the same way it behaves when IP addresses are specified that are not actual Elasticsearch hosts.

**Steps to reproduce**:
 1. download ES5.0.0alpha5 tarball and extract
 2.  cd to directory and run:

```
ES_JAVA_OPTS="-Xms512m -Xmx512m" ./bin/elasticsearch -Ecluster.name=oom_test  -Enetwork.host=_site_   -Ediscovery.zen.ping.unicast.hosts=oom.elastic.com
```

Compare with the behavior when this non-ES host IP is specified:
```
ES_JAVA_OPTS="-Xms512m -Xmx512m" ./bin/elasticsearch -Ecluster.name=oom_test  -Enetwork.host=_site_ -Ediscovery.zen.minimum_master_nodes=1   -Ediscovery.zen.ping.unicast.hosts=1.2.3.4
```
**Provide logs (if relevant)**:

```
[2016-09-02 08:18:17,972][INFO ][node                     ] [] initializing ...
[2016-09-02 08:18:18,030][INFO ][env                      ] [_zjSJnZ] using [1] data paths, mounts [[/ (/dev/disk2)]], net usable_space [873.3gb], net total_space [1tb], spins? [unknown], types [hfs]
[2016-09-02 08:18:18,030][INFO ][env                      ] [_zjSJnZ] heap size [494.9mb], compressed ordinary object pointers [true]
[2016-09-02 08:18:18,040][INFO ][node                     ] [_zjSJnZ] node name [_zjSJnZ] derived from node ID; set [node.name] to override
[2016-09-02 08:18:18,042][INFO ][node                     ] [_zjSJnZ] version[5.0.0-alpha5], pid[2161], build[d327dd4/2016-08-04T08:59:39.568Z], OS[Mac OS X/10.11.6/x86_64], JVM[Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/1.8.0_31/25.31-b07]
[2016-09-02 08:18:18,663][INFO ][io.netty.util.internal.PlatformDependent] Your platform does not provide complete low-level API for accessing direct buffers reliably. Unless explicitly requested, heap buffer will always be preferred to avoid potential system unstability.
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [aggs-matrix-stats]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [ingest-common]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [lang-expression]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [lang-groovy]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [lang-mustache]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [lang-painless]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [percolator]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [reindex]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [transport-netty3]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] loaded module [transport-netty4]
[2016-09-02 08:18:18,666][INFO ][plugins                  ] [_zjSJnZ] no plugins loaded
java.lang.OutOfMemoryError: Java heap space
Dumping heap to java_pid2161.hprof ...
Heap dump file created [277176889 bytes in 1.559 secs]

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "main"
```

**Possibly related to**: [https://github.com/elastic/elasticsearch/issues/14441](https://github.com/elastic/elasticsearch/issues/14441)

**Comment by clintongormley** [comment]
@jasontedor could you take a look at this please?

**Comment by jasontedor** [comment]
This reproduces on 5.0.0-alpha5 but not master for me. I think that it might be due to the massive stack trace from Guice that is attempted to be logged and I think our recent switch to Log4j 2 resolves the `OutOfMemoryError`. Can you please verify @loren?

**Comment by jasontedor** [comment]
I've verified this is addressed by #20083, which was an incremental improvement made while working on Log4j 2.

**Comment by jasontedor** [comment]
Closed by #20083
